### ISSUE TITLE
💡: the extension should ignore files outside the open workspace

### ISSUE THREAD

**Issue by heaths** [streetsidesoftware/vscode-spell-checker#3581]
### Problem

In VSCode, even if we get our repo cspell-clean, if we F12 into other files outside the open workspace e.g., under `~/.rustup/toolchains/` for downloaded `std` source, the **Problems** pane is populated with a bunch of misspellings we have no control over.

### Solution

At least for the VSCode extension, any files outside the current workspace folder(s) should be ignored. More generally, any directories outside the current working directory could be ignored.

### Alternatives

If this is the purpose of the `--root` (usage docs aren't entirely clear how it's used - presumably only for `--relative`), that could also be configurable in the configuration file e.g., in `.vscode/cspell.json` we could add `"root": ".."`.

### Additional Context

Once misspellings are added to the **Problems** pane, even if the file outside the workspace with the unknown words is closed, those problems hang around and are very distracting. We can filter out "Infos" from the **Problem** pane, but that may also hide some "infos" we want to see from other extensions.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by swaldhoer** [comment]
I have set `"spellCheckOnlyWorkspaceFiles": true` in `.vscode/cspell.json `, but VS Code tells me `Property spellCheckOnlyWorkspaceFiles is not allowed.`
 
And when I check files outside the workspace, cspell still shows errors that should be ignored.

The folder is opened as workspace, not as a single file.
 
Version: streetsidesoftware.code-spell-checker@3.0.1

**Comment by Jason3S** [comment]
> I have set `"spellCheckOnlyWorkspaceFiles": true` in `.vscode/cspell.json `, but VS Code tells me `Property spellCheckOnlyWorkspaceFiles is not allowed.`   And when I check files outside the workspace, cspell still shows errors that should be ignored.
>
> The folder is opened as workspace, not as a single file.   Version: streetsidesoftware.code-spell-checker@3.0.1

Sorry for the confusion. This is not a CSpell CLI setting, it is a setting only used by the extension. Not all extension settings can be added to a `cspell.json` file.

Add it to `.vscode/settings.json`:

<img width="460" alt="image" src="https://github.com/user-attachments/assets/e93fa930-0cd7-44d6-89d1-7f03e4551e99">

Or use `files` in the `cspell.json` file:

**`.vscode/cspell.json`**
```jsonc
{
    "globRoot": "..", // glob root is needed because this file is inside `.vscode/`
    "files": ["/**"]
}
```

**Comment by swaldhoer** [comment]
Ah okay, thanks for the clarification. I will test tomorrow.

A side question: so it is possible to configure some things in in different configuration files and the configurations are getting merged?

**Comment by Jason3S** [comment]
> A side question: so it is possible to configure some things in in different configuration files and the configurations are getting merged?

CSpell CLI configuration can be merged. VS Code settings have their own rules: [Settings Precedence](https://code.visualstudio.com/docs/getstarted/settings#_settings-precedence).

The extension uses the CSpell library to do the spell checking. The CSpell library searches up the directory tree for the nearest CSpell configuration file. Settings found in the CSpell configuration are merged with the settings from VS Code.

It is possible to use the `import` setting to explicitly import CSpell configuration.

Here is an example `import` to import config from `<workspace root>/project-config/cspell.config.yaml` and the [German dictionary](https://www.npmjs.com/package/@cspell/dict-de-de) (needs to be installed separately).

**`.vscode/cspell.json`**
```jsonc
{
  "import": ["../project-config/cspell.config.yaml", "@cspell/dict-de-de"]
}
```

**`.vscode/settings.json`**
```jsonc
"cSpell.import": ["${workspaceRoot}/project-config/cspell.config.yaml", "@cspell/dict-de-de"] 
// ${workspaceRoot} (see below) is needed because the extension doesn't known the
// location of the `settings.json` file. 
``` 

Special Values in `.vscode/settings.json`:
- `${workapceRoot}` - The first workspace folder.
- `${workspaceFolder}` - Imports will be relative to the current workspace folder
- `${workspaceFolder:<name>}` - Where `<name>` is the name of the workspace folder.

**Comment by swaldhoer** [comment]
> Or use `files` in the `cspell.json` file:
> 
> **`.vscode/cspell.json`**
> 
> ```js
> {
>     "globRoot": "..", // glob root is needed because this file is inside `.vscode/`
>     "files": ["/**"]
> }
> ```

I tried this and it does not work.
Should I provide screenshots?

**Comment by tylerlaprade** [comment]
The mentioned setting does not work. None of the approaches above prevent the extensions from complaining about the spellings in `settings.json`.

**Comment by Jason3S** [comment]
@tylerlaprade,


VS Code opens files in the background. When it opens the files, it triggers the spell checker.

There are two work arounds:

1. Use custom decorations: `cSpell.useCustomDecorations`
   <img width="554" alt="image" src="https://github.com/user-attachments/assets/109a0b85-3864-4793-a4d5-1631926c5769">
   The spell checker will handle rendering issues and files opened by VS Code in the background will not show up.
1. The `Spell: Toggle Show Spelling Issues`, `Spell: Show Spelling Issues`, and `Spell: Hide Spelling Issues` commands can be used to show/hide spelling issues. Using these command will clean up and left over issues in the problem pane.
   <img width="380" alt="image" src="https://github.com/user-attachments/assets/a4bbd743-6e39-457b-a41f-304e189856f2">

**Comment by Jason3S** [comment]
I'm going to close this for now. There are two different issues being raised here.

1. Only spell checking files within the workspace.
2. Showing spelling issues for files that are opened in the background by VS Code.

Issue 1. should be working. If it is not, please open a new issue with your configuration.
I hope to have a better solution for Issue 2. Please realize it was caused by a change in behavior by VS Code.
